cpp: introduce the allocator

tomaszkapela opened the pull request:
- Add the persistent memory allocator.
- Modifiy makefiles to enable custom libsdtc++ and libc++
  installations.
- Add basic tests. The container tests. Then need a slightly
  modified version of libc++ to be able to compile.

The allocator introduces the persistency_type typedef, which is
used by the modified libc++ implementation for appropriate
container metadata handling.

This commit does not include Windows related projects for the
unit tests. That is because VS2015 does not handle allocators
with fancy pointers correctly. The debug version of the containers
will most certainly segfault on these tests. It is probable
that VS2017 will handle this a lot better.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/1545)
<!-- Reviewable:end -->

GBuella commented on src/include/libpmemobj++/detail/persistent_ptr_base.hpp:
```diff
@@ -0,0 +1,354 @@
+	element_type *
+	get() const noexcept
+	{
+		auto ret = this->oid.pool_uuid_lo ==
```
How about 
if () {} else {}
?

tomaszkapela replied:
Done.

GBuella commented on src/include/libpmemobj++/detail/persistent_ptr_base.hpp:
```diff
@@ -0,0 +1,354 @@
+			? reinterpret_cast<element_type *>(oid.off)
+			: static_cast<element_type *>(
+				  pmemobj_direct(this->oid));
+		return ret;
```
Why assign to a temp variable?

tomaszkapela replied:
Done.

GBuella commented on src/include/libpmemobj++/detail/specialization.hpp:
```diff
@@ -125,6 +125,12 @@ struct sp_array_access {
 
+template <>
+struct sp_array_access<void> {
+	typedef struct does_not_exist {
```
https://cdn.meme.am/cache/instances/folder661/500x/56497661.jpg

tomaszkapela replied:
But at least this produces a nice compilation error ;)

GBuella commented on src/include/libpmemobj++/persistent_ptr.hpp:
```diff
@@ -217,78 +143,23 @@ class persistent_ptr {
 	typename nvml::detail::sp_array_access<T>::type
 	operator[](std::ptrdiff_t i) const noexcept
 	{
 		assert(i >= 0 && (i < nvml::detail::sp_extent<T>::value ||
```
This necessarily has to be part of introducing the allocator?

tomaszkapela replied:
The thing is, the allocator does not make sense if it does not work nicely with standard library containers and containers need pointers. One special kind is a `void` pointer which has rules of its own. I probably could separate this into two commits, is that what you reeeeaaaly want?

GBuella replied:
No, what I really want is an explanation like this.

marcinslusarz commented:
> Linux tests are available, however they are not compiled on travis
> yet, because a non-standard, modified version of libc++ is needed.

We could include custom libc++ in docker image and use in travis build.

tomaszkapela commented:
>We could include custom libc++ in docker image and use in travis build.

Baby steps.

marcinslusarz commented:
Reviewed 27 of 97 files at r1, 3 of 7 files at r2, 2 of 4 files at r3, 1 of 1 files at r4.
Review status: 33 of 96 files reviewed at latest revision, 15 unresolved discussions.

---

*[src/include/libpmemobj++/allocator.hpp, line 107 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_Z7Xo-eE7hjcfXVBvK:-K_Z7Xo-eE7hjcfXVBvL:bic5jz) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/include/libpmemobj++/allocator.hpp#L107)):*
> ```cpp
> 	address(reference r)
> 	{
> 		/* pointer to volatile problem */
> ```

volatile problem?

---

*[src/include/libpmemobj++/allocator.hpp, line 289 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_Z85javQnuxxJd2qi6:-K_Z85jb6_XH40DzQnT4:bcyh1xj) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/include/libpmemobj++/allocator.hpp#L289)):*
> ```cpp
> 			sizeof(value_type) * cnt, detail::type_num<T>());
> 
> 		return ptr;
> ```

return pmemobj_tx_alloc(); ?

---

*[src/include/libpmemobj++/allocator.hpp, line 308 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_Z8GPbC0FeQGwk85a5:-K_Z8GPbC0FeQGwk85a6:b-xwtf7p) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/include/libpmemobj++/allocator.hpp#L308)):*
> ```cpp
> 
> 		if (pmemobj_tx_free(*p.raw_ptr()) != 0)
> 			throw transaction_alloc_error(
> ```

transaction_free_error

---

*[src/include/libpmemobj++/allocator.hpp, line 395 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_Z8PxnSa7Pq-b6onKK:-K_Z8PxnSa7Pq-b6onKL:b-7ropln) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/include/libpmemobj++/allocator.hpp#L395)):*
> ```cpp
> 			pmemobj_tx_alloc(1 /* void size */ * cnt, 0);
> 
> 		return ptr;
> ```

ditto

---

*[src/include/libpmemobj++/allocator.hpp, line 414 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_Z8UWkS2LGjM3VG93a:-K_Z8UWkS2LGjM3VG93b:b-7ropln) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/include/libpmemobj++/allocator.hpp#L414)):*
> ```cpp
> 
> 		if (pmemobj_tx_free(p.raw()) != 0)
> 			throw transaction_alloc_error(
> ```

ditto

---

*[src/include/libpmemobj++/README.md, line 37 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZFv24McOf7DN2KYNW:-K_ZFv24McOf7DN2KYNX:bvpz2lh) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/include/libpmemobj++/README.md#L37)):*
> ```Markdown
> consistent across subsequent executions or compilations of user applications.
> 
> The C++ bindings have an experimental, standard copmliant memory allocator which
> ```

compliant

"have"? maybe "implement"?

---

*[src/test/Makefile.inc, line 242 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZI5E4pPSLtlCPJ1Bc:-K_ZI5E4pPSLtlCPJ1Bd:bqwb7n6) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/test/Makefile.inc#L242)):*
> ```Makefile
> 
> ifeq ($(COMPILE_LANG), cpp)
> CXX=clang++
> ```

spaces around = please

---

*[src/test/Makefile.inc, line 258 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZILaraQYf3RU1HiWh:-K_ZILaraQYf3RU1HiWi:b-goxidp) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/test/Makefile.inc#L258)):*
> ```Makefile
> 
> ifeq ($(COMPILE_LANG), cpp)
> CC = $(GCC_BINDIR)/gcc
> ```

Why do you define CC here, but not when LLVM is used?

---

*[src/test/obj_cpp_allocator/Makefile, line 2 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZIgxUZJL0xXzM8gSe:-K_ZIgxUZJL0xXzM8gSf:bb6ht2g) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/test/obj_cpp_allocator/Makefile#L2)):*
> ```Makefile
> #
> # Copyright 2015-2016, Intel Corporation
> ```

2016

---

*[src/test/obj_cpp_allocator/obj_cpp_allocator.cpp, line 2 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZIZ7U1NyB5xbb9KcC:-K_ZIZ7U1NyB5xbb9KcD:b-fayb2z) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/test/obj_cpp_allocator/obj_cpp_allocator.cpp#L2)):*
> ```cpp
> /*
>  * Copyright 2015-2016, Intel Corporation
> ```

2016, Please fix this in other files.

---

*[src/test/obj_cpp_multimap/TEST0, line 37 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZKDCSxuxgG7w8a26R:-K_ZKDCSxuxgG7w8a26S:b2qg71q) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/test/obj_cpp_multimap/TEST0#L37)):*
> ```
> export UNITTEST_NUM=0
> 
> export VALGRIND_OPTS="--suppressions=../memcheck-libunwind.supp"
> ```

Maybe we should add it to default suppression list?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1545)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Well, if tests are never run on Travis we don't know whether allocator works :P

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1545#-:-K_ZOewZmwXv7QRgmd21:b6qjemx)*
<!-- Sent from Reviewable.io -->

GBuella commented:
Review status: 33 of 96 files reviewed at latest revision, 16 unresolved discussions.

---

*[src/test/Makefile.inc, line 258 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZyjTlgE8euTbf7zE4:-K_ZyjTlgE8euTbf7zE5:b-pmvb05) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/test/Makefile.inc#L258)):*
> ```Makefile
> 
> ifeq ($(COMPILE_LANG), cpp)
> CC = $(GCC_BINDIR)/gcc
> ```

I think silently overwriting CC and/or CXX is wrong.
We should use the user supplied toolchains, and one wants to be clever and helpful, can issue a warning in the makefile, that says something like "You seem to use the wrong toolchain".
Or perhaps better, stop the build altogether.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1545)*
<!-- Sent from Reviewable.io -->

tomaszkapela commented:
Review status: 33 of 96 files reviewed at latest revision, 16 unresolved discussions.

---

*[src/include/libpmemobj++/allocator.hpp, line 107 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_Z7Xo-eE7hjcfXVBvK:-K__FR_nSbl7Xl9WlrMp:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/include/libpmemobj++/allocator.hpp#L107)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

volatile problem?
</blockquote></details>

Done.

---

*[src/include/libpmemobj++/allocator.hpp, line 289 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_Z85javQnuxxJd2qi6:-K_ZT5YuWtbHEK0wy7Wp:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/include/libpmemobj++/allocator.hpp#L289)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

return pmemobj_tx_alloc(); ?
</blockquote></details>

Done.

---

*[src/include/libpmemobj++/allocator.hpp, line 308 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_Z8GPbC0FeQGwk85a5:-K__IkIyTqJpMZJgEwZc:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/include/libpmemobj++/allocator.hpp#L308)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

transaction_free_error
</blockquote></details>

Done.

---

*[src/include/libpmemobj++/allocator.hpp, line 395 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_Z8PxnSa7Pq-b6onKK:-K__ByxJodZQKx79xcY0:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/include/libpmemobj++/allocator.hpp#L395)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[src/include/libpmemobj++/allocator.hpp, line 414 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_Z8UWkS2LGjM3VG93a:-K__IleuLWquhrSFiS7J:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/include/libpmemobj++/allocator.hpp#L414)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[src/include/libpmemobj++/README.md, line 37 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZFv24McOf7DN2KYNW:-K__B5RTtHXIK4hhB_pj:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/include/libpmemobj++/README.md#L37)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

compliant

"have"? maybe "implement"?
</blockquote></details>

Done.

---

*[src/test/Makefile.inc, line 242 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZI5E4pPSLtlCPJ1Bc:-K__Abs2Y1o_9KxpSZFy:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/test/Makefile.inc#L242)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

spaces around = please
</blockquote></details>

Done.

---

*[src/test/Makefile.inc, line 258 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZyjTlgE8euTbf7zE4:-K__AdhgMi2vnbIb4tlq:bok3tx8) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/test/Makefile.inc#L258)):*
<details><summary><i>Previously, GBuella (Gábor Buella) wrote…</i></summary><blockquote>

I think silently overwriting CC and/or CXX is wrong.
We should use the user supplied toolchains, and one wants to be clever and helpful, can issue a warning in the makefile, that says something like "You seem to use the wrong toolchain".
Or perhaps better, stop the build altogether.
</blockquote></details>

I believe what @gbuella wrote is right. I removed the silent overwrite of *CC/CXX* and added a note to README.md

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1545)*
<!-- Sent from Reviewable.io -->

tomaszkapela commented:
Review status: 26 of 98 files reviewed at latest revision, 16 unresolved discussions.

---

*[src/test/obj_cpp_allocator/Makefile, line 2 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZIgxUZJL0xXzM8gSe:-K_d-4ZPm1Z9Mijo5yF-:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/test/obj_cpp_allocator/Makefile#L2)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

2016
</blockquote></details>

Done.

---

*[src/test/obj_cpp_allocator/obj_cpp_allocator.cpp, line 2 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZIZ7U1NyB5xbb9KcC:-K_d-4qPEtXVDutMll-_:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/test/obj_cpp_allocator/obj_cpp_allocator.cpp#L2)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

2016, Please fix this in other files.
</blockquote></details>

Done.

---

*[src/test/obj_cpp_multimap/TEST0, line 37 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZKDCSxuxgG7w8a26R:-K_dI50sJqWskxQNDnps:bu5fasq) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/test/obj_cpp_multimap/TEST0#L37)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Maybe we should add it to default suppression list?
</blockquote></details>

It's only useful for C++ tests using libc++, but I think there is no harm in adding it as a default suppression. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1545)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
*[src/include/libpmemobj++/README.md, line 37 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZFv24McOf7DN2KYNW:-K_dTiqWUh5rkX-OfkxQ:b-1fys1e) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/include/libpmemobj++/README.md#L37)):*
<details><summary><i>Previously, tomaszkapela (Tomasz Kapela) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

still a typo in "compliant"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1545)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Reviewed 2 of 7 files at r2, 73 of 74 files at r5.
Review status: 97 of 98 files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1545)*
<!-- Sent from Reviewable.io -->

tomaszkapela commented:
Review status: 97 of 98 files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---

*[src/include/libpmemobj++/README.md, line 37 at r4](https://reviewable.io:443/reviews/pmem/nvml/1545#-K_ZFv24McOf7DN2KYNW:-K_dYi91WpYoMU0mDc04:b-2hgsrm) ([raw file](https://github.com/pmem/nvml/blob/ba4bc6065979a28adaa98fd3b75795c5d2842713/src/include/libpmemobj++/README.md#L37)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

still a typo in "compliant"
</blockquote></details>

Sorry, somehow missed that. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1545)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1545#-:-K_duFz3cNEPaIRMoKoq:bnfp4nl)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1545)*
<!-- Sent from Reviewable.io -->
